lib/using_restricted_array.rb

```diff
@@ -4,60 +4,204 @@
+      return true
+    end
+    i += 1
+    end
+  return false
 end
 
+# DELETE AFTER TESTING
+puts "testing search method"
+test_arr = [188, 45, 129]
+pp test_arr
+p search(test_arr, 3, 129)
+puts ""
+
+############################################################################
+
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); requires n steps or n operations where n is the length of the array
+# Space complexity: O(n); requires n steps or n operations where n is the length of the array
 def find_largest(array, length)
-  raise NotImplementedError
+  if length == 0
+    return nil
+  end
+
+  i = 0
+  largest_value = array[0]
+  while i < length
+    if array[i] > largest_value
+      largest_value = array[i]
+    end
+    i += 1
+  end
+  return largest_value
 end
 
+# DELETE AFTER TESTING
+puts "testing find_largest method"
+test_arr = [188, 45, 189]
+pp test_arr
+p find_largest(test_arr, 3)
+puts ""
+############################################################################
+
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); requires n steps or n operations where n is the length of the array
+# Space complexity: O(n); requires n steps or n operations where n is the length of the array
 def find_smallest(array, length)
-  raise NotImplementedError
+  if length == 0
+    return nil
+  end
+
+  i = 0
+  smallest_value = array[0]
+  while i < length
+    if array[i] < smallest_value
+      smallest_value = array[i]
+    end
+    i += 1
+  end
+  return smallest_value
 end
 
+# DELETE AFTER TESTING
+puts "testing smallest method"
+test_arr = [188, 45, 129]
+pp test_arr
+p find_smallest(test_arr, 3)
+puts ""
+
+############################################################################
+
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); requires n steps or n operations where n is the length of the array
+# Space complexity: O(n); requires n steps or n operations where n is the length of the array
 def reverse(array, length)
-  raise NotImplementedError
+  if length <= 1
+    return nil
+  end
+
+  i = 0
+  j = (length - 1)
+
+  while i < j
+    temp = array[i]
+    array[i] = array[j]
+    array[j] = temp
+    i += 1
+    j -= 1
+  end
+  return array
 end
 
+puts "testing reverse method"
+test_arr = [188, 45, 129]
+pp test_arr
+p reverse(test_arr, 3)
+puts ""
+
+############################################################################
+
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(logn)
+# Space complexity: O(logn)
 def binary_search(array, length, value_to_find)
```

Comment:
👍 However your space complexity is O(1) since you are not building any other data structure.  Your method doesn't increase in memory usage if the input is bigger.